rs/orchestrator/src/catch_up_package_provider.rs

```diff
@@ -210,14 +223,78 @@ impl CatchUpPackageProvider {
     // caller.
     async fn fetch_catch_up_package(
         &self,
-        url: Url,
+        node_id: &NodeId,
+        url: String,
         param: Option<CatchUpPackageParam>,
     ) -> Option<pb::CatchUpPackage> {
-        Agent::new_with_client(self.client.clone(), url, Sender::Anonymous)
-            .query_cup_endpoint(param)
+        let body = Bytes::from(
+            param
+                .and_then(|param| serde_cbor::to_vec(&param).ok())
+                .unwrap_or_default(),
+        );
+
+        let client_config = self
+            .crypto_tls_config
+            .client_config(*node_id, self.registry.get_latest_version())
+            .map_err(|e| warn!(self.logger, "Failed to create tls client config: {:?}", e))
+            .ok()?;
+
+        let https = HttpsConnectorBuilder::new()
+            .with_tls_config(client_config)
+            .https_only()
+            .enable_all_versions()
+            .build();
+
+        let client = Client::builder(TokioExecutor::new())
+            .pool_idle_timeout(tokio::time::Duration::from_secs(600))
+            .pool_max_idle_per_host(1)
+            .build::<_, Full<Bytes>>(https);
+
+        let req = tokio::time::timeout(
+            tokio::time::Duration::from_secs(10),
+            client.request(
+                Request::builder()
+                    .method(Method::POST)
+                    .header("content-type", "application/cbor")
+                    .uri(url)
+                    .body(Full::from(body))
+                    .map_err(|e| warn!(self.logger, "Failed to create request: {:?}", e))
+                    .ok()?,
+            ),
+        );
+
+        let res = req
             .await
+            .map_err(|e| warn!(self.logger, "Querying CUP endpoint timed out: {:?}", e))
+            .ok()?
             .map_err(|e| warn!(self.logger, "Failed to query CUP endpoint: {:?}", e))
+            .ok()?;
+
+        let bytes = res
+            .into_body()
+            .collect()
+            .await
+            .map_err(|e| {
+                warn!(
+                    self.logger,
+                    "Failed to convert the response body to bytes: {:?}", e
+                )
+            })
             .ok()?
+            .to_bytes();
+
+        if bytes.is_empty() {
+            None
+        } else {
+            pb::CatchUpPackage::decode(&bytes[..])
+                .map_err(|e| {
+                    warn!(
+                        self.logger,
+                        "Failed to deserialize CUP from protobuf, got: {:?} - error {:?}", bytes, e
```

Comment:
This may be a lot of bytes, I'm not sure printing them would add any value
```suggestion
                        "Failed to deserialize CUP from protobuf: {:?}", e
```